app/frontend/src/components/AvatarInput.test.tsx

```diff
@@ -0,0 +1,165 @@
+      expect(
+        screen.getByAltText(getAvatarLabel(NAME)).getAttribute("src")
+      ).toMatch(new RegExp(MOCK_THUMB));
+    });
+  });
+
+  it("cancels when cancel button pressed and doesn't submit key", async () => {
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), MOCK_FILE);
+
+    expect(await screen.findByLabelText(CANCEL_UPLOAD)).toBeVisible();
+
+    userEvent.click(screen.getByLabelText(CANCEL_UPLOAD));
+
+    await waitFor(() => {
+      expect(uploadFileMock).not.toHaveBeenCalled();
+    });
+
+    userEvent.click(screen.getByRole("button", { name: SUBMIT }));
+
+    await waitFor(() => {
+      expect(submitForm).toHaveBeenCalledWith({ avatarInput: "" });
+    });
+    expect(screen.getByAltText(getAvatarLabel(NAME))).toHaveProperty(
+      "src",
+      MOCK_INITIAL_SRC
+    );
+  });
+
+  it(`uploading, confirming, then uploading again, then cancelling, goes back to the first upload instead of the original`, async () => {
+    const OTHER_MOCK_FILE = new File([], "firstImage.jpg");
+    uploadFileMock.mockResolvedValueOnce({
+      file: OTHER_MOCK_FILE,
+      filename: OTHER_MOCK_FILE.name,
+      key: "firstKey",
+      thumbnail_url: "thumb0.jpg",
+      full_url: "full0.jpg",
+    });
+
+    //first upload and confirm
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), OTHER_MOCK_FILE);
+    expect(await screen.findByLabelText(CONFIRM_UPLOAD)).toBeVisible();
+    userEvent.click(screen.getByLabelText(CONFIRM_UPLOAD));
+
+    await waitFor(() => {
+      expect(uploadFileMock).toHaveBeenCalled();
+    });
+    expect(
+      screen.getByAltText(getAvatarLabel(NAME)).getAttribute("src")
+    ).toMatch(/thumb0.jpg/);
+
+    //2nd upload and cancel
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), MOCK_FILE);
+    expect(await screen.findByLabelText(CANCEL_UPLOAD)).toBeVisible();
+    userEvent.click(screen.getByLabelText(CANCEL_UPLOAD));
+    expect(
+      (await screen.findByAltText(getAvatarLabel(NAME))).getAttribute("src")
+    ).toMatch(/thumb0.jpg/);
+
+    //submit
+    userEvent.click(screen.getByRole("button", { name: SUBMIT }));
+
+    await waitFor(() => {
+      expect(submitForm).toHaveBeenCalledWith({ avatarInput: "firstKey" });
+    });
+  });
+
+  it("doesn't submit without confirming/cancelling", async () => {
+    userEvent.upload(screen.getByLabelText(SELECT_AN_IMAGE), MOCK_FILE);
+
+    expect(await screen.findByLabelText(CONFIRM_UPLOAD)).toBeVisible();
+
+    userEvent.click(screen.getByRole("button", { name: SUBMIT }));
+
+    expect(await screen.findByText(UPLOAD_PENDING_ERROR)).toBeVisible();
+    expect(submitForm).not.toHaveBeenCalled();
+  });
+
+  it.skip("displays an error for an invalid file", async () => {
+    userEvent.upload(
+      screen.getByLabelText(SELECT_AN_IMAGE),
+      new File([new Blob(undefined)], "")
+    );
+    expect(await screen.findByText(INVALID_FILE)).toBeVisible();
+  });
```

Comment:
No matter what I tried, couldn't make an invalid file to test this....

Reply:
I got it working by making `readAsDataURL` emit an `error` event on all `FileReader` instances for this test specifically, which I guess is what happens when there's an error as these seem to use the event callback pattern rather than throwing an error directly, so I think it's fine to do? More interested in hitting the error handling path rather than how it's achieved here...

```suggestion
  it("displays an error for an invalid file", async () => {
    jest.spyOn(FileReader.prototype, "readAsDataURL").mockImplementation(() => {
      FileReader.prototype.dispatchEvent(new Event("error"));
    });
    userEvent.upload(
      screen.getByLabelText(SELECT_AN_IMAGE),
      new File([new Blob(undefined)], "")
    );
    expect(await screen.findByText(INVALID_FILE)).toBeVisible();
  });
```